File: zig/src/tools/eval_dispatcher.zig
//! eval-dispatcher — language-agnostic dispatcher pilot eval runner.
//!
//! Reads JSONL ops from stdin, runs each through XmlToolDispatcher or
//! NativeToolDispatcher, writes one canonical-JSON response per data-returning
//! op to stdout. Mirrors eval-tools/src/bin/eval-dispatcher.rs.

const std = @import("std");
const zeroclaw = @import("zeroclaw");
const dispatcher = zeroclaw.runtime.agent.dispatcher;

const EvalError = error{InvalidScenario};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const input = try std.io.getStdIn().readToEndAlloc(allocator, 16 * 1024 * 1024);
    defer allocator.free(input);

    const stdout = std.io.getStdOut().writer();
    var lines = std.mem.splitScalar(u8, input, '\n');
    while (lines.next()) |raw_line| {
        const line = std.mem.trim(u8, raw_line, " \t\r\n");
        if (line.len == 0) continue;
        try runOp(allocator, line, stdout);
    }
}

fn runOp(allocator: std.mem.Allocator, line: []const u8, writer: anytype) !void {
    var parsed = try std.json.parseFromSlice(std.json.Value, allocator, line, .{});
    defer parsed.deinit();
    const op = getString(parsed.value, "op") orelse return EvalError.InvalidScenario;
    const dispatcher_kind = getString(parsed.value, "dispatcher") orelse return EvalError.InvalidScenario;

    if (std.mem.eql(u8, op, "parse_response")) {
        try runParseResponse(allocator, parsed.value, dispatcher_kind, writer);
    } else if (std.mem.eql(u8, op, "format_results")) {
        try runFormatResults(allocator, parsed.value, dispatcher_kind, writer);
    } else if (std.mem.eql(u8, op, "should_send_tool_specs")) {
        try runShouldSendToolSpecs(dispatcher_kind, writer);
    } else {
        return EvalError.InvalidScenario;
    }
}

fn runParseResponse(
    allocator: std.mem.Allocator,
    op_value: std.json.Value,
    dispatcher_kind: []const u8,
    writer: anytype,
) !void {
    const response_value = getField(op_value, "response") orelse return EvalError.InvalidScenario;
    const text = getOptionalString(response_value, "text");
    const tool_calls_value = getField(response_value, "tool_calls");

    var tool_calls = std.ArrayList(dispatcher.ToolCall).init(allocator);
    defer tool_calls.deinit();
    if (tool_calls_value) |tcv| {
        if (tcv == .array) {
            for (tcv.array.items) |tc| {
                if (tc != .object) return EvalError.InvalidScenario;
                try tool_calls.append(.{
                    .id = getString(tc, "id") orelse return EvalError.InvalidScenario,
                    .name = getString(tc, "name") orelse return EvalError.InvalidScenario,
                    .arguments = getString(tc, "arguments") orelse return EvalError.InvalidScenario,
                });
            }
        }
    }

    const response = dispatcher.ChatResponse{
        .text = text,
        .tool_calls = tool_calls.items,
    };

    var result = try invokeParseResponse(allocator, dispatcher_kind, response);
    defer result.deinit(allocator);

    try writer.writeAll("{\"op\":\"parse_response\",\"result\":");
    try writeParseResult(writer, result);
    try writer.writeAll("}\n");
}

fn runFormatResults(
    allocator: std.mem.Allocator,
    op_value: std.json.Value,
    dispatcher_kind: []const u8,
    writer: anytype,
) !void {
    const results_value = getField(op_value, "results") orelse return EvalError.InvalidScenario;
    if (results_value != .array) return EvalError.InvalidScenario;

    var results = std.ArrayList(dispatcher.ToolExecutionResult).init(allocator);
    defer results.deinit();
    for (results_value.array.items) |entry| {
        if (entry != .object) return EvalError.InvalidScenario;
        const success_v = getField(entry, "success") orelse return EvalError.InvalidScenario;
        if (success_v != .bool) return EvalError.InvalidScenario;
        try results.append(.{
            .name = getString(entry, "name") orelse return EvalError.InvalidScenario,
            .output = getString(entry, "output") orelse return EvalError.InvalidScenario,
            .success = success_v.bool,
            .tool_call_id = getOptionalString(entry, "tool_call_id"),
        });
    }

    var msg = try invokeFormatResults(allocator, dispatcher_kind, results.items);
    defer msg.deinit(allocator);

    try writer.writeAll("{\"op\":\"format_results\",\"result\":");
    try writeConversationMessage(writer, msg);
    try writer.writeAll("}\n");
}

fn runShouldSendToolSpecs(dispatcher_kind: []const u8, writer: anytype) !void {
    const result = blk: {
        if (std.mem.eql(u8, dispatcher_kind, "xml")) {
            var x = dispatcher.XmlToolDispatcher{};
            break :blk x.dispatcher().shouldSendToolSpecs();
        } else if (std.mem.eql(u8, dispatcher_kind, "native")) {
            var n = dispatcher.NativeToolDispatcher{};
            break :blk n.dispatcher().shouldSendToolSpecs();
        } else return EvalError.InvalidScenario;
    };
    try writer.writeAll("{\"op\":\"should_send_tool_specs\",\"result\":");
    try writer.writeAll(if (result) "true" else "false");
    try writer.writeAll("}\n");
}

fn invokeParseResponse(
    allocator: std.mem.Allocator,
    dispatcher_kind: []const u8,
    response: dispatcher.ChatResponse,
) !dispatcher.ParseResult {
    if (std.mem.eql(u8, dispatcher_kind, "xml")) {
        var x = dispatcher.XmlToolDispatcher{};
        return x.dispatcher().parseResponse(allocator, response);
    } else if (std.mem.eql(u8, dispatcher_kind, "native")) {
        var n = dispatcher.NativeToolDispatcher{};
        return n.dispatcher().parseResponse(allocator, response);
    }
    return EvalError.InvalidScenario;
}

fn invokeFormatResults(
    allocator: std.mem.Allocator,
    dispatcher_kind: []const u8,
    results: []const dispatcher.ToolExecutionResult,
) !dispatcher.ConversationMessage {
    if (std.mem.eql(u8, dispatcher_kind, "xml")) {
        var x = dispatcher.XmlToolDispatcher{};
        return x.dispatcher().formatResults(allocator, results);
    } else if (std.mem.eql(u8, dispatcher_kind, "native")) {
        var n = dispatcher.NativeToolDispatcher{};
        return n.dispatcher().formatResults(allocator, results);
    }
    return EvalError.InvalidScenario;
}

// ─── JSON output helpers ──────────────────────────────────────────────────

fn writeParseResult(writer: anytype, result: dispatcher.ParseResult) !void {
    try writer.writeAll("{\"text\":");
    try std.json.stringify(result.text, .{}, writer);
    try writer.writeAll(",\"calls\":[");
    for (result.calls, 0..) |call, i| {
        if (i > 0) try writer.writeByte(',');
        try writer.writeAll("{\"name\":");
        try std.json.stringify(call.name, .{}, writer);
        try writer.writeAll(",\"arguments\":");
        try std.json.stringify(call.arguments, .{}, writer);
        try writer.writeAll(",\"tool_call_id\":");
        if (call.tool_call_id) |id| {
            try std.json.stringify(id, .{}, writer);
        } else {
            try writer.writeAll("null");
        }
        try writer.writeAll("}");
    }
    try writer.writeAll("]}");
}

fn writeConversationMessage(writer: anytype, msg: dispatcher.ConversationMessage) !void {
    // Mirrors serde tag="type", content="data" on
    // zeroclaw_api::provider::ConversationMessage.
    switch (msg) {
        .chat => |chat| {
            try writer.writeAll("{\"type\":\"Chat\",\"data\":{\"role\":");
            try std.json.stringify(chat.role, .{}, writer);
            try writer.writeAll(",\"content\":");
            try std.json.stringify(chat.content, .{}, writer);
            try writer.writeAll("}}");
        },
        .tool_results => |results| {
            try writer.writeAll("{\"type\":\"ToolResults\",\"data\":[");
            for (results, 0..) |r, i| {
                if (i > 0) try writer.writeByte(',');
                try writer.writeAll("{\"tool_call_id\":");
                try std.json.stringify(r.tool_call_id, .{}, writer);
                try writer.writeAll(",\"content\":");
                try std.json.stringify(r.content, .{}, writer);
                try writer.writeAll("}");
            }
            try writer.writeAll("]}");
        },
    }
}

// ─── JSON input helpers ───────────────────────────────────────────────────

fn getField(value: std.json.Value, key: []const u8) ?std.json.Value {
    if (value != .object) return null;
    return value.object.get(key);
}

fn getString(value: std.json.Value, key: []const u8) ?[]const u8 {
    const field = getField(value, key) orelse return null;
    if (field != .string) return null;
    return field.string;
}

fn getOptionalString(value: std.json.Value, key: []const u8) ?[]const u8 {
    const field = getField(value, key) orelse return null;
    return switch (field) {
        .null => null,
        .string => |inner| inner,
        else => null,
    };
}

File: eval-tools/src/bin/eval-dispatcher.rs
//! eval-dispatcher — language-agnostic dispatcher pilot eval runner.
//!
//! Reads JSONL ops from stdin, runs each through XmlToolDispatcher or
//! NativeToolDispatcher, writes one canonical-JSON response per data-returning
//! op to stdout. Mirrors zig/src/tools/eval_dispatcher.zig.

use std::io::{self, Read, Write};

use anyhow::{Context, Result};
use serde_json::Value;
use zeroclaw_providers::{ChatResponse, ToolCall};
use zeroclaw_runtime::agent::dispatcher::{
    NativeToolDispatcher, ParsedToolCall, ToolDispatcher, ToolExecutionResult, XmlToolDispatcher,
};

fn main() -> Result<()> {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;
    let mut output = Vec::new();

    for raw_line in input.lines() {
        let line = raw_line.trim();
        if line.is_empty() {
            continue;
        }
        let op_value: Value = serde_json::from_str(line).context("invalid scenario JSON")?;
        let op = op_value
            .get("op")
            .and_then(Value::as_str)
            .context("op missing")?;
        let dispatcher_kind = op_value
            .get("dispatcher")
            .and_then(Value::as_str)
            .context("dispatcher missing")?;
        let dispatcher = make_dispatcher(dispatcher_kind)?;

        match op {
            "parse_response" => {
                let response = chat_response_from_json(&op_value)?;
                let (text, calls) = dispatcher.parse_response(&response);
                let calls_json: Vec<Value> = calls.iter().map(parsed_call_to_json).collect();
                let result = serde_json::json!({"text": text, "calls": calls_json});
                write_result(&mut output, op, result);
            }
            "format_results" => {
                let results = tool_execution_results_from_json(&op_value)?;
                let msg = dispatcher.format_results(&results);
                write_result(&mut output, op, serde_json::to_value(msg)?);
            }
            "should_send_tool_specs" => {
                write_result(
                    &mut output,
                    op,
                    serde_json::json!(dispatcher.should_send_tool_specs()),
                );
            }
            _ => anyhow::bail!("unknown op: {op}"),
        }
    }

    io::stdout().lock().write_all(&output)?;
    Ok(())
}

fn make_dispatcher(kind: &str) -> Result<Box<dyn ToolDispatcher>> {
    match kind {
        "xml" => Ok(Box::new(XmlToolDispatcher)),
        "native" => Ok(Box::new(NativeToolDispatcher)),
        _ => anyhow::bail!("unknown dispatcher kind: {kind}"),
    }
}

fn chat_response_from_json(op: &Value) -> Result<ChatResponse> {
    let response = op.get("response").context("response missing")?;
    let text = response
        .get("text")
        .and_then(Value::as_str)
        .map(String::from);
    let tool_calls: Vec<ToolCall> = if let Some(arr) =
        response.get("tool_calls").and_then(Value::as_array)
    {
        arr.iter()
            .map(|tc| -> Result<ToolCall> {
                Ok(ToolCall {
                    id: tc
                        .get("id")
                        .and_then(Value::as_str)
                        .context("tool_call.id missing")?
                        .to_string(),
                    name: tc
                        .get("name")
                        .and_then(Value::as_str)
                        .context("tool_call.name missing")?
                        .to_string(),
                    arguments: tc
                        .get("arguments")
                        .and_then(Value::as_str)
                        .context("tool_call.arguments missing")?
                        .to_string(),
                })
            })
            .collect::<Result<Vec<_>>>()?
    } else {
        Vec::new()
    };
    Ok(ChatResponse {
        text,
        tool_calls,
        usage: None,
        reasoning_content: None,
    })
}

fn tool_execution_results_from_json(op: &Value) -> Result<Vec<ToolExecutionResult>> {
    let arr = op
        .get("results")
        .and_then(Value::as_array)
        .context("results missing")?;
    arr.iter()
        .map(|r| -> Result<ToolExecutionResult> {
            Ok(ToolExecutionResult {
                name: r
                    .get("name")
                    .and_then(Value::as_str)
                    .context("name missing")?
                    .to_string(),
                output: r
                    .get("output")
                    .and_then(Value::as_str)
                    .context("output missing")?
                    .to_string(),
                success: r
                    .get("success")
                    .and_then(Value::as_bool)
                    .context("success missing")?,
                tool_call_id: r
                    .get("tool_call_id")
                    .and_then(Value::as_str)
                    .map(String::from),
            })
        })
        .collect()
}

fn parsed_call_to_json(call: &ParsedToolCall) -> Value {
    serde_json::json!({
        "name": call.name,
        "arguments": call.arguments,
        "tool_call_id": call.tool_call_id,
    })
}

fn write_result(output: &mut Vec<u8>, op: &str, result: Value) {
    let value = serde_json::json!({"op": op, "result": result});
    output.extend_from_slice(canonical_dump(&value).as_bytes());
    output.push(b'\n');
}

fn canonical_dump(value: &Value) -> String {
    match value {
        Value::Null => "null".to_string(),
        Value::Bool(b) => b.to_string(),
        Value::Number(n) => n.to_string(),
        Value::String(s) => serde_json::to_string(s).unwrap(),
        Value::Array(arr) => {
            let items: Vec<String> = arr.iter().map(canonical_dump).collect();
            format!("[{}]", items.join(","))
        }
        Value::Object(obj) => {
            let mut keys: Vec<&String> = obj.keys().collect();
            keys.sort();
            let items: Vec<String> = keys
                .iter()
                .map(|k| {
                    format!(
                        "{}:{}",
                        serde_json::to_string(k).unwrap(),
                        canonical_dump(&obj[*k])
                    )
                })
                .collect();
            format!("{{{}}}", items.join(","))
        }
    }
}
